ansible_builder/main.py

```diff
@@ -49,18 +54,39 @@ def build_command(self):
             self.build_context
         ]
 
+    def run_in_container(self, command, **kwargs):
+        wrapped_command = [
+            self.container_runtime, 'run',
+            '--rm', self.tag, '/bin/bash', '-c'
+        ] + [' '.join(command)]
+        return run_command(wrapped_command, **kwargs)
+
     def build(self):
         self.containerfile.create_folder_copy_files()
         self.containerfile.prepare_prepended_steps()
-        self.containerfile.prepare_introspection_steps()
         self.containerfile.prepare_galaxy_steps()
+        self.containerfile.prepare_introspection_steps()
         print(MessageColors.OK + 'Writing partial Containerfile without collection requirements' + MessageColors.ENDC)
         self.containerfile.write()
-        run_command(self.build_command)
-        self.containerfile.prepare_pip_steps()
-        print(MessageColors.OK + 'Rewriting Containerfile to capture collection requirements' + MessageColors.ENDC)
-        self.containerfile.prepare_system_steps()
+        rc, output = run_command(self.build_command, capture_output=True)
+        collection_data = ansible_builder.introspect.parse_introspect_output('\n'.join(output))
+        if collection_data is None:
+            # In case the build has already been ran once, a cached layer may be used for introspect
+            rc, output = self.run_in_container(['introspect'], capture_output=True)
+            collection_data = ansible_builder.introspect.parse_introspect_output('\n'.join(output))
+        if collection_data.get('system'):
+            rc, output = self.run_in_container(
+                [
+                    'pip3', 'install', 'bindep;',
+                    'echo', '----begin_bindep_output----;',
+                    'bindep', '-b', '-f', '/build/{0}'.format(BINDEP_COMBINED)],
```

Comment:
Here, I both install and run bindep in a new container. This keeps bindep out of the python environment in the image, see https://github.com/ansible/ansible-builder/pull/76#issuecomment-674635447

Reply:
For downstream we'll need to have bindep in the base image, so we should probably do that upstream too.